llvm8: support c++ in cross case

matthewbauer opened the pull request:
this adds libc++ to the LLVM cross, giving us access to the full
Nixpkgs set. This requires 4 stages of wrapped compilers:

- Clang with no libraries
- Clang with just compiler-rt
- Clang with Libc, and compiler-rt
- Clang with Libc++, Libc, and compiler-rt

<!-- Nixpkgs has a lot of new incoming Pull Requests, but not enough people to review this constant stream. Even if you aren't a committer, we would appreciate reviews of other PRs, especially simple ones like package updates. Just testing the relevant package/service and leaving a comment saying what you tested, how you tested it and whether it worked would be great. List of open PRs: <https://github.com/NixOS/nixpkgs/pulls>, for more about reviewing contributions: <https://hydra.nixos.org/job/nixpkgs/trunk/manual/latest/download/1/nixpkgs/manual.html#sec-reviewing-contributions>. Reviewing isn't mandatory, but it would help out a lot and reduce the average time-to-merge for all of us. Thanks a lot if you do! -->
###### Motivation for this change


###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Assured whether relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

Test:

```
with import ./default.nix {
  crossSystem = (import ./lib).systems.examples.aarch64-multiplatform-musl // {useLLVM = true;};
};
let
cpp = builtins.toFile "test.c++" ''
#include <iostream>

int main() {
  std::cout << "Hello world!" << std::endl;
}
'';
testdrv = { callPackage, stdenv, llvmPackages_8 }: stdenv.mkDerivation
    {
        name = "cross-test";
        version = "0.1";
        unpackPhase = "true";
        buildInputs = [llvmPackages_8.libunwind];
        buildPhase = ''
            $CXX -lunwind -o test ${cpp}
        '';
        installPhase = ''
            mkdir -p $out/bin
            cp test $out/bin/test
            ${stdenv.hostPlatform.emulator buildPackages} $out/bin/test
        '';
    };
in callPackage testdrv {}
```

---

matthewbauer committed:
llvm8: support c++ in cross case

this adds libc++ to the LLVM cross, giving us access to the full
Nixpkgs set. This requires 4 stages of wrapped compilers:

- Clang with no libraries
- Clang with just compiler-rt
- Clang with Libc, and compiler-rt
- Clang with Libc++, Libc, and compiler-rt

Ericson2314 commented on pkgs/development/compilers/llvm/8/default.nix:
```diff
@@ -78,48 +78,78 @@ let
+    # pulled in. As a consequence, it is very quick to build different
+    # targets provided by LLVM and we can also build for what GCC
+    # doesn’t support like LLVM. Probably we should move to some other
+    # file.
```
If we have `libc`, `libcxx`, `librt`, `libunwind`` attributes in cc-wrapper, I think most of this logic could move there even though some it is clang-specific.

matthewbauer replied:
You would still need the 4 ccs though right?

Ericson2314 replied:
Perhaps but it could arise organically like:
```nix
// all-packages.nix
libc = ...;
librt = ...;
```
```
// llvm default.nix
xxx = ... (stdenvNoLlibs.override {
  foo = libraries.noNull; 
}) ...;
```

Ericson2314 committed:
clang 8, lldClangStdenv: Get LLVM's unwind working properly

Backport the `--unwindlib` flag, so it don't try to use libgcc on Linux
for hysterical reasons. The alternative of passing `-lunwind` always
would mess up C.

Ericson2314 commented:
Pushed a WIP thing with unwind to my own repo, but it didn't work right off the bat.
